File: src/main/java/com/GoldenMine/parser/parser/SentenceMultiDataParser.java
package com.GoldenMine.parser.parser;

import com.GoldenMine.parser.Code;
import com.GoldenMine.parser.ParseContext;
import com.GoldenMine.parser.Sentence;
import com.GoldenMine.parser.predicatespecific.IPredicateSpecific;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SentenceMultiDataParser implements IParser {
    private IParser parser;

    public SentenceMultiDataParser(IParser parser) {
        this.parser = parser;
    }

    @Override
    public void parse(Code code, ParseContext context, int index) {
        if (parser != null)
            parser.parse(code, context, index);
        if(!context.isNoParse()) {
            List<Sentence> sentences = context.getSentences();

            //for (int sentenceIndex = ; sentenceIndex < sentences.size(); sentenceIndex++)
            {
                Sentence sentence = sentences.get(sentences.size() - 1);
                IPredicateSpecific specific = sentence.getSpecific();


                if (specific != null) {
                    List<String> others = specific.getOthers();
                    System.out.println("not null specific " + index + ", " + others.size());

                    int overlap = 0;
                    Integer[] multiIndex = new Integer[others.size() + 1];
                    multiIndex[0] = index;
                    for (int i = 1; i < multiIndex.length; i++) {
                        multiIndex[i] = -1;
                    }
                    int set = 0;

                    for (int line = index + 1; line < code.getSourceCode().size(); line++) {
                        ParseContext parseContext = code.getSourceCode().get(line);
                        List<Sentence> innerSentences = parseContext.getSentences();
                        String source = parseContext.getSource();

                        for (int innerSentenceIndex = 0; innerSentenceIndex < innerSentences.size(); innerSentenceIndex++) {
                            Sentence innerSentence = innerSentences.get(innerSentenceIndex);

                            if (specific == innerSentence.getSpecific()) {
                                overlap++;
                            }
                        }
                        if(parseContext.isNoParse()) {
                            int containsSource = others.indexOf(source);
                            System.out.println(source + ", " + others + ", " + containsSource);
                            if (containsSource >= 0) {
                                if (overlap == 0) {
                                    multiIndex[containsSource + 1] = line;
                                    set++;
                                    if (set >= others.size()) {
                                        break;
                                    }
                                } else {
                                    if (containsSource == others.size() - 1) {
                                        overlap--;
                                    }
                                }
                            }
                        }

                        System.out.println(line + ", " + overlap);
                    }

                    //System.out.println(multiIndex[0] + ", " + multiIndex[1]);

                    IPredicateSpecific.MultiProcessData data = new IPredicateSpecific.MultiProcessData(Arrays.asList(multiIndex));

                    sentence.setMultiProcessData(data);
                    for (int i = 1; i < multiIndex.length; i++) {
                        //System.out.println(multiIndex[i]==null);
                        int value = multiIndex[i];
                        if (value != -1) {
                            ParseContext secondary = code.getSourceCode().get(value);
//                            List<Sentence> secSentences = secondary.getSentences();
//
//                            for(int senIndex = 0; senIndex < secSentences.size(); senIndex++) {
//                                Sentence secSentence = secSentences.get(senIndex);
//
//                            }
                            if(secondary.isNoParse()) {
                                Sentence defaultSentence = new Sentence(null, sentence.get서술어(), sentence.getSpecific());
                                defaultSentence.setMultiProcessData(data);
                                secondary.getSentences().add(defaultSentence);
                            }
                        }
                    }
                }
            }
        }
    }
}

File: src/main/java/kr/goldenmine/parser/parser/SentenceMultiDataParser.kt
package kr.goldenmine.parser.parser

import kr.goldenmine.parser.Code
import kr.goldenmine.parser.ParseContext
import kr.goldenmine.parser.Sentence
import kr.goldenmine.parser.predicatespecific.IPredicateSpecific

class SentenceMultiDataParser(private val parser: IParser?) : IParser {

    override fun parse(code: Code, parseContext: ParseContext, index: Int, debug: Boolean, metadata: List<Any>?) {
        parser?.parse(code, parseContext, index, debug, metadata)
        if (!parseContext.isNoParse) {
            val sentences = parseContext.sentences

            //for (int sentenceIndex = ; sentenceIndex < sentences.size(); sentenceIndex++)
            run {
                val sentence = sentences[sentences.size - 1]
                val specific = sentence.specific

                if (specific != null) {
                    val others = specific.others
                    //System.out.println("not null specific " + index + ", " + others.size());

                    var overlap = 0
                    //val multiIndex = arrayOfNulls<Int>(others.size + 1)
                    val multiIndex = IntArray(others.size + 1)
                    multiIndex[0] = index
                    for (i in 1 until multiIndex.size) {
                        multiIndex[i] = -1
                    }
                    var set = 0

                    for (line in index + 1 until code.sourceCode.size) {
                        val parseContext = code.sourceCode[line]
                        val innerSentences = parseContext.sentences
                        val source = parseContext.source

                        for (innerSentenceIndex in innerSentences.indices) {
                            val innerSentence = innerSentences[innerSentenceIndex]

                            if (specific === innerSentence.specific) {
                                overlap++
                            }
                        }
                        if (parseContext.isNoParse) {
                            val containsSource = others.indexOf(source)
                            //System.out.println(source + ", " + others + ", " + containsSource);
                            if (containsSource >= 0) {
                                if (overlap == 0) {
                                    multiIndex[containsSource + 1] = line
                                    set++
                                    if (set >= others.size) {
                                        break
                                    }
                                } else {
                                    if (containsSource == others.size - 1) {
                                        overlap--
                                    }
                                }
                            }
                        }

                        //System.out.println(line + ", " + overlap);
                    }

                    //println("${multiIndex[0]} ${multiIndex[1]} ")

                    val data = IPredicateSpecific.MultiProcessData(multiIndex.toCollection(ArrayList()))

                    sentence.multiProcessData = data
                    for (i in 1 until multiIndex.size) {
                        //System.out.println(multiIndex[i]==null);
                        val value = multiIndex[i]
                        if (value != -1) {
                            val secondary = code.sourceCode[value]
                            //                            List<Sentence> secSentences = secondary.getSentences();
                            //
                            //                            for(int senIndex = 0; senIndex < secSentences.size(); senIndex++) {
                            //                                Sentence secSentence = secSentences.get(senIndex);
                            //
                            //                            }
                            if (secondary.isNoParse) {
                                val defaultSentence = Sentence(HashMap(), sentence.서술어, sentence.specific)
                                defaultSentence.multiProcessData = data
                                secondary.sentences.add(defaultSentence)
                            }
                        }
                    }
                }
            }
        }
    }
}
